### ISSUE TITLE
Region written stats is abnormal

### ISSUE THREAD

**Issue by tabokie** [tikv/tikv#9796]
### What version of TiKV are you using?

Versions that I know of: 4.0.10, 5.0.0-rc

### What did you expect?

Value of region written bytes make sense, and is relatively stable.

### What did happened?

In this case (5.0.0-rc), service is running normally, but region written bytes reach multiple PBs

![image](https://user-images.githubusercontent.com/27005812/110889028-58ff7800-8328-11eb-8172-205c64f1b8ac.png)

In this case (4.0.10), two down servers worsen the abnormal state

![image](https://user-images.githubusercontent.com/27005812/110888901-1a69bd80-8328-11eb-88a5-48796a563744.png)

**Comment by tabokie** [comment]
In one of the cases, the `tikv_region_written_bytes_sum` value is 1.8e19, it's likely a u64 overflow, possibly caused by `delta_bytes` in `ApplyContext`.

**Comment by BusyJay** [comment]
Because the average written bytes keeps growing, and it's calculated by
```
let written_bytes = current_written_bytes - last_written_bytes;
last_written_bytes = current_written_bytes;
```
. If it's underflow, then `current_written_bytes` keeps decreasing for more than 5 hours, which doesn't seem reasonable.

**Comment by tabokie** [comment]
@BusyJay Are you referring to the second screenshot? I recheck the metrics, it seems like something wrong with grafana, the linear graph is a display error:

![image](https://user-images.githubusercontent.com/27005812/117647067-168ad800-b1bf-11eb-90f0-40b0b43d85c1.png)

**Comment by BusyJay** [comment]
So it seems like the metrics is not updated for 5 hours, and then hit underflow?

**Comment by cosven** [comment]
Will the region scheduling affected by these stats?

/severity moderate

**Comment by BusyJay** [comment]
If metrics is not updated for 5 hours, then either the clock of the prometheus machine jumps or tikv/pd thread is stuck for 5 hours. If the latter is the case, and only pd thread is stuck, then scheduling will not work for the time; if tikv is stuck, then it should be completely out of service.

**Comment by tabokie** [comment]
It seems there are two separate cases here:
1) Metrics slowly increasing, the value of `tikv_region_written_bytes_sum` has not reach u64::max
2) `tikv_region_written_bytes_sum` instantly jumping to u64::max, suspect u64 underflow

Those two cases can be randomly triggered when restarting tikv nodes, or killing one tikv node and restart it after a while.

Assigning this to @MrCroxx for further investigation.

**Comment by MrCroxx** [comment]
For problem 1, the bug locates in `pub fn heartbeat_pd` of `component/raftstore/src/store/peer.rs`.

When a region is created, the leader will spawn 3 `heartbeat_pd` task. 1 for notify its leadership, 2 for notify that one of its peers has caught up (3 replicas). Normally, `heartbeat_pd` will spawn a task and let `pd_scheduler` schedule it, but when `approximate_xxx` is initially 0,  `heartbeat_pd` will spawn a task with `approximate_xxx` data, then pass it to `split_check_scheduler`, then use callback to let `pd_scheduler` to scheduler it. It will lead to the 3 tasks in wrong order as expected.

The bad case comes up when there is a `ApplyRes` message between the messages which spawn the 3 tasks. `ApplyRes` will trigger `fn post_apply` which advance `last_written_xxx`. If the tasks are in wrong order, the `written_xxx` will be `0 -> some unzero value -> 0`, and will cause `last_written_xxx` underflow.

I logged the related data and found a sample:
![image](https://user-images.githubusercontent.com/22407295/120259155-ffbf3900-c2c5-11eb-8936-6d83e86b6510.png)

**Comment by MrCroxx** [comment]
[This](https://github.com/tikv/tikv/pull/10119) pull request fixed this bug and was picked to release-5.0 . Need to pick it to release-4.0 .
